### PR TITLE
Support Unix domain sockets

### PR DESCRIPTION
**Pull request by djones6**
## Description

Depends on https://github.com/IBM-Swift/Kitura-NIO/pull/187 / https://github.com/IBM-Swift/Kitura-net/pull/296

Adds the ability to listen on a Unix domain socket, rather than an INET (TCP) port.

## How Has This Been Tested?

Tests added.  Updated KituraTest to run existing tests over Unix domain sockets.

## Checklist:

- [x] If applicable, I have updated the documentation accordingly.
- [x] If applicable, I have added tests to cover my changes.

### CODE REVIEW

File: Sources/Kitura/Kitura.swift [hunk 1]
```diff
@@ -64,13 +64,48 @@ public class Kitura {
...
         server.sslConfig = sslConfig?.config
         server.keepAliveState = keepAliveState
         server.allowPortReuse = allowPortReuse
         serverLock.lock()
-        httpServersAndPorts.append((server: server, port: port))
+        if let port = port {
+            httpServersAndPorts.append((server: server, port: port))
+        } else if let socketPath = socketPath {
+            httpServersAndUnixSocketPaths.append((server: server, socketPath: socketPath))
+        }
```
Comment (ianpartridge) [inline Sources/Kitura/Kitura.swift]: What happens if both `port` and `socketPath` are nil? Silent failure. I know this is impossible today but let's add a final `else{}` here for safety.

File: Tests/KituraTests/KituraTest.swift [hunk 2]
```diff
@@ -63,41 +92,61 @@ class KituraTest: XCTestCase {
...
+            }
         }
     }
 
     func doPerformServerTest(router: ServerDelegate, timeout: TimeInterval, line: Int, asyncTasks: [(XCTestExpectation) -> Void]) {
 
-        guard let port = startServer(router: router) else {
-            XCTFail("Error starting server. useSSL:\(useSSL)")
-            return
+        if useUnixSocket {
```
Comment (ianpartridge) [inline Tests/KituraTests/KituraTest.swift]: `self.`? And throughout the PR.. I think it's a good practice.

File: Tests/KituraTests/KituraTest.swift [hunk 3]
```diff
@@ -189,6 +302,34 @@ class KituraTest: XCTestCase {
     func expectation(line: Int, index: Int) -> XCTestExpectation {
         return self.expectation(description: "\(type(of: self)):\(line)[\(index)](ssl:\(useSSL))")
     }
+
+    // Generates a unique temporary file path suitable for use as a Unix domain socket.
+    // On Linux, a path is returned within /tmp
+    // On MacOS, a path is returned within /var/folders
+    func uniqueTemporaryFilePath() -> String {
```
Comment (ianpartridge) [inline Tests/KituraTests/KituraTest.swift]: Can we use Foundation on Linux too?

Comment (djones6) [reply Tests/KituraTests/KituraTest.swift]: FileManager.temporaryDirectory wasn't implemented in Swift 4.0.3

File: Tests/KituraTests/KituraTest.swift (hunk 2 above)
Comment (djones6) [reply Tests/KituraTests/KituraTest.swift]: I changed this to use a private enum: https://github.com/IBM-Swift/Kitura/pull/1434/commits/026f967f56e4300f21cf5e740dd59c3a5d1e4615

File: Sources/Kitura/Kitura.swift (hunk 1 above)
Comment (djones6) [reply Sources/Kitura/Kitura.swift]: I changed this to use a private enum: https://github.com/IBM-Swift/Kitura/pull/1434/commits/026f967f56e4300f21cf5e740dd59c3a5d1e4615

**Comment by djones6** [comment]
@ianpartridge Because NIO 2 requires Swift 5, it'll be cleaner to merge #1436 first and then rebase this.

**Pull request merged**